extensions/amp-ad/0.1/amp-ad-xorigin-iframe-handler.js

```diff
@@ -125,60 +122,78 @@ export class AmpAdXOriginIframeHandler {
-    if (this.baseInstance_.config
-        && this.baseInstance_.config.renderStartImplemented) {
-      // If support render-start, create a race between render-start no-content
-      this.adResponsePromise_ = listenForOncePromise(this.iframe,
+    // Calculate render-start and no-content signal. These signals are mutually
+    // exlcusive. Whichever arrives first wins.
+    let renderStartResolve;
+    const renderStartPromise = new Promise(resolve => {
+      renderStartResolve = resolve;
+    });
+    let noContentResolve;
+    const noContentPromise = new Promise(resolve => {
+      noContentResolve = resolve;
+    });
+    listenForOncePromise(this.iframe,
         ['render-start', 'no-content'], true).then(info => {
           const data = info.data;
           if (data.type == 'render-start') {
             this.renderStart_(info);
+            renderStartResolve();
           } else {
             this.noContent_();
+            noContentResolve();
           }
         });
-    } else {
-      // If NOT support render-start, listen to bootstrap-loaded no-content
-      // respectively
-      this.adResponsePromise_ = listenForOncePromise(this.iframe,
-        'bootstrap-loaded', true);
-      listenForOncePromise(this.iframe, 'no-content', true)
-          .then(() => this.noContent_());
-    }
 
     if (opt_isA4A) {
       // A4A writes creative frame directly to page therefore does not expect
       // post message to unset visibility hidden
       this.element_.appendChild(this.iframe);
       return Promise.resolve();
     }
-    // Set iframe initially hidden which will be removed on load event +
-    // post message.
+
+    // Set iframe initially hidden which will be removed on render-start or
+    // load, whichever is earlier.
     setStyle(this.iframe, 'visibility', 'hidden');
     this.element_.appendChild(this.iframe);
+    const visibilityPromise = Promise.race([
+      renderStartPromise,
+      iframeLoadPromise,
+    ]).then(() => {
+      if (this.iframe) {
+        setStyle(this.iframe, 'visibility', '');
+        if (this.baseInstance_.emitLifecycleEvent) {
+          this.baseInstance_.emitLifecycleEvent('adSlotUnhidden');
+        }
+      }
+    });
 
-    return timerFor(this.baseInstance_.win).timeoutPromise(TIMEOUT_VALUE,
-        this.adResponsePromise_,
-        'timeout waiting for ad response').catch(e => {
-          this.noContent_();
-          user().warn('AMP-AD', e);
-        }).then(() => {
-          if (this.iframe) {
-            setStyle(this.iframe, 'visibility', '');
-            if (this.baseInstance_.emitLifecycleEvent) {
-              this.baseInstance_.emitLifecycleEvent('adSlotUnhidden');
-            }
-          }
-        });
+    // The actual ad load is eariliest of iframe.onload event and no-content.
+    const adLoadPromise = Promise.race([
+      iframeLoadPromise,
+      noContentPromise,
+    ]);
+    return adLoadPromise.then(() => {
```

Comment:
i thought we discussed that layoutCallback should be resolved purely on iframeLoad?
This way we separate concepts so that:
- iframe onload controls layoutCallback, which controls resource scheduling
- render-start and no-content controls UI experience